File: ddb/sdk/typescript/examples/breakpoint.ts
import { newClient } from "./config.js";

const client = newClient();
try {
  const snapshot = (await client.call("DebuggerService.GetSnapshot", {
    sections: ["SNAPSHOT_SECTION_TOPOLOGY", "SNAPSHOT_SECTION_EXECUTION"],
  })).snapshot;
  const thread = snapshot?.threads?.find((candidate) => candidate.state === "THREAD_STATE_STOPPED");
  if (!thread?.sessionId || !thread.location?.path || !thread.location.line) {
    throw new Error("a stopped source-backed thread is required");
  }
  const admitted = await client.call("DebuggerControlService.CreateBreakpoint", {
    target: { session: { sessionId: thread.sessionId } },
    breakpoint: {
      source: { source: thread.location.path, line: thread.location.line },
      enabled: true,
    },
  });
  const created = await client.waitOperation(admitted.operation?.operationId ?? "");
  const breakpointId = created.result?.breakpoint?.breakpointId;
  if (!breakpointId) throw new Error("DDB omitted the breakpoint result");
  console.log("created", breakpointId);

  const deleted = await client.call("DebuggerControlService.DeleteBreakpoint", {
    breakpointId,
    target: { session: { sessionId: thread.sessionId } },
  });
  await client.waitOperation(deleted.operation?.operationId ?? "");
} finally {
  client.close();
}

File: ddb/sdk/python/examples/breakpoint.py
from _config import new_client


with new_client() as client:
    snapshot = client.get_snapshot(
        {"sections": ["SNAPSHOT_SECTION_TOPOLOGY", "SNAPSHOT_SECTION_EXECUTION"]}
    ).get("snapshot", {})
    thread = next(
        (item for item in snapshot.get("threads", []) if item.get("state") == "THREAD_STATE_STOPPED"),
        None,
    )
    location = thread.get("location", {}) if thread else {}
    if not thread or not thread.get("sessionId") or not location.get("path") or not location.get("line"):
        raise RuntimeError("a stopped source-backed thread is required")
    admitted = client.create_breakpoint(
        {
            "target": {"session": {"sessionId": thread["sessionId"]}},
            "breakpoint": {
                "source": {"source": location["path"], "line": location["line"]},
                "enabled": True,
            },
        }
    )
    operation = client.wait_operation(admitted.get("operation", {}).get("operationId", ""))
    breakpoint_id = operation.get("result", {}).get("breakpoint", {}).get("breakpointId")
    if not breakpoint_id:
        raise RuntimeError("DDB omitted the breakpoint result")
    print("created", breakpoint_id)
    deleted = client.delete_breakpoint(
        {
            "breakpointId": breakpoint_id,
            "target": {"session": {"sessionId": thread["sessionId"]}},
        }
    )
    client.wait_operation(deleted.get("operation", {}).get("operationId", ""))
